src/Network/Transport/TCP.hs

```diff
@@ -629,12 +663,189 @@ apiNewEndPoint transport =
     resolveMulticastGroupError =
       TransportError ResolveMulticastGroupUnsupported "Multicast not supported"
 
+-- | Specification of input queueing for a 'LocalEndPoint'
+data QPolicy prio = QPolicy {
+    -- | Determine a buffer constructor given an EndPointAddress.
+    --   Parametricity ensures the buffer is empty.
+    qpolicyBuffer :: forall t . EndPointAddress -> Buffer t
+    -- | Determine a priority for each EndPointAddress. Values which compare
+    --   lower give higher priority.
+  , qpolicyPriority :: EndPointAddress -> prio
+  }
+
+-- | Round-robin queuing with unbounded queues for everybody.
+roundRobinUnbounded :: QPolicy ()
+roundRobinUnbounded = QPolicy {
+    -- Unbounded buffers for everybody.
+    qpolicyBuffer = const emptyUnbounded
+    -- Equal priority for everybody.
+  , qpolicyPriority = const ()
+  }
+
+-- | Round-robin queuing with bounded queues for everybody.
+--   Minimum valid bound is 1.
+roundRobinBounded :: Int -> QPolicy ()
+roundRobinBounded bound
+  | bound < 1 = throw $ userError "roundRobinBounded minimum bound is 1"
+  | otherwise = QPolicy {
+        qpolicyBuffer = const (emptyBounded (bound - 1))
+      , qpolicyPriority = const ()
+      }
+
+-- | State of input queue and buffers for a 'LocalEndPoint'.
+data QDisc prio = QDisc {
+    -- | Ordered Events from 'RemoteEndPoint's.
+    --   All keys are elements in qdiscQueue.
+    qdiscBuffers :: Map EndPointAddress (NonEmptyBuffer Event)
+    -- | All values in this priority queue are keys in qdiscBuffers.
+  , qdiscQueue :: PQ.MinPQueue prio EndPointAddress
+    -- | What to do in case we try to read from an empty queue.
+    --   Typically this will be 'retry'. After a LocalEndPoint is closed, it
+    --   will be an STM which delivers the EndPointClosed event.
+  , qdiscIdle :: STM Event
+    -- | A queueing policy. It's in STM so that it can be changed over time.
+  , qdiscPolicy :: STM (QPolicy prio)
+  }
+
+instance Ord prio => Show (QDisc prio) where
+  show qdisc = concat [
```

Comment:
:minor: I'm not sure if using Show instance for pretty-printing is a great practice but works here..